### ISSUE TITLE
Repo consolidation for Governance and ousd contracts

### ISSUE THREAD

**Issue by franckc** [OriginProtocol/origin-dollar#1097]
Having governance contracts in a separate repo from ousd makes the testing harder. We'd like to consolidate all the contracts under the ousd repo.

TODO
 - [ ] Move governance contracts over to the OUSD repo
 - [ ] Rewrite governance brownie tests to hardhat

Caveat: governance and ousd contracts use different versions for
 - Solidity
 - Open Zeppelin contracts 
We'd like to preserve that so that we don't have to re-deploy all the contracts.

**Comment by micahalcorn** [comment]
Is the second part of this the same as https://github.com/OriginProtocol/ousd-governance/issues/292 ?

**Comment by ivanzhelyazkov** [comment]
I have some architectural questions about the move of the governance contracts to the OUSD repo.
Imo there are 3 ways to do that:
1. Merge the governance and OUSD repos in one monorepo
2. Import the governance repo as a submodule in the OUSD repo
3. Copy the contracts from the gov repo only and keep them up-to-date in the OUSD repo

Imo the best way to do this is the second way, since we wouldn't need to maintain two versions of the contracts, and it won't be such a big organisational change.
What do you think on this ?

**Comment by franckc** [comment]
How about just moving the contracts from the governance repo to the origin-dollar repo? Is that what you mean by option 3?

**Comment by sparrowDom** [comment]
@ivanzhelyazkov I think the best way to go would be Nr.1. The goals we want to achieve are: 
- have all the contracts compiled using hardhat instead of brownie
- have the same deployment tooling (hardhat) when updating / deploying governance & OUSD contracts
- still have the ability to use brownie console as it is used right now

I don't think submodules are a good approach because:
- with submodules we would still have 2 repos and 2 pull requests when code changes impact OUSD & OUSD governance contracts
- OUSD governance will not be a self sustained module anymore. Compiling and deploying will be taken care by OUSD repo tooling

Feel free to disagree of course as to why other approaches make more sense

**Comment by ivanzhelyazkov** [comment]
> @ivanzhelyazkov I think the best way to go would be Nr.1. But instead of creating a new repo, I would just consolidate everything in OUSD repo. The goals we want to achieve are:
> 
> * have all the contracts compiled using hardhat instead of brownie
> * have the same deployment tooling (hardhat) when updating / deploying governance & OUSD contracts
> * still have the ability to use brownie console as it is used right now
> 
> I don't think submodules are a good approach because:
> 
> * with submodules we would still have 2 repos and 2 pull requests when code changes impact OUSD & OUSD governance contracts
> * OUSD governance will not be a self sustained module anymore. Compiling and deploying will be taken care by OUSD repo tooling
> 
> Feel free to disagree of course as to why other approaches make more sense

For me it makes sense to consolidate everything in one repo, only issue I see there is tracking the git history of the changes to OUSD and OUSD-governance will become more difficult.

**Comment by sparrowDom** [comment]
Yeah that is a good point. I guess for some transitional time we will have to refer to the repo we discontinue to check for historical changed. Which isn't too bad IMHO.

**Comment by ivanzhelyazkov** [comment]
An update on this issue - I've gone forward with the consolidation of OUSD governance contracts in the Origin-dollar repo and have hit a problem which will impact the resulting contracts in the repo. There are five contracts which have identical names in OUSD and OUSD governance. I've compiled the files successfully without changing the names, but hit several blocks while trying to run the tests.
Another solution will be to alter the contracts names or merge some of the contracts which are (almost) indentical.

Posting my suggestions here, feel free to comment.

Contract collisions between OUSD and OUSD-governance:

- Governor (from Openzeppelin@4.6.0 package)
- Governable (OUSD governance)
- Timelock (OUSD governance)
- InitializeGovernedUpgradeabilityProxy (OUSD governance)
- MockOGV (OUSD governance)

Proposed changes for solving the contract collisions between OUSD and OUSD-governance:

* Governor (from Openzeppelin@4.6.0 package):
 Proposed solution:
 	- change name to OriginGovernor, and change in tests everywhere
 Notes:
 	- Since OZ package contract name can't be changed and both contracts are quite different, this is an option currently, even though it requires a lot of changes to the tests

* InitializeGovernedUpgradeabilityProxy (from ousd-governance)
 Proposed solution:
 	- change governance contract name to InitializeGovernedUpgradeabilityProxyGovernance
 Notes:
 	-- Very similar to ousd contract, but we have an address dependency which is from 4.4.1 in the OUSD case, and 4.6.0 in the OUSD-governance case - the difference there is the isAddress check, which in 4.4.1 uses extcodesize assembly and 4.6.0 uses address.code.length

* Timelock (from ousd-governance)
 Proposed Solution:
 	- Replace OUSD-governance Timelock with TimelockGovernance
 Notes:
 	- Contracts are different, ousd-governance uses TimelockController from OZ package

* Governable (from ousd-governance)
 Proposed Solution:
 	- Replace OUSD-governance contract with the OUSD one in governance/ dir
 	- Replace governance contracts which import Governable to OUSD one 
 Notes:
 	- Only differences between the two contracts is the abstract modifier and the license.

* MockOGV (from ousd-governance)
 Proposed Solutions:
 	- Replace OUSD-governance contract with the OUSD one
 	- Change MockOGV to MockOGVGovernance
 Notes:
 	- Despite the MockOGV from OUSD repo using OZ 4.4.1 ERC-20 standard and MockOGV from OUSD-governance repo using OZ 4.6.0 ERC-20 standard (which have some implementation differences), the contracts are used for testing and not part of the live deployment
 	- Leaning a bit more toward using just one contract

**Comment by shahthepro** [comment]
> MockOGV (from ousd-governance)

For MockOGV, we only use these mock contracts in unit tests. So having a storage layout change in that contract shouldn't be a problem at all.

> Governor (from Openzeppelin@4.6.0 package):
> change name to OriginGovernor, and change in tests everywhere

Fair enough

> InitializeGovernedUpgradeabilityProxy (from ousd-governance)

There seems to be [one breaking change in the OZ Address contract](https://github.com/OpenZeppelin/openzeppelin-contracts/releases/tag/v4.5.0), between ^4.4.0 and ^4.6.0. So the proposed solution sounds good to me. One thing to note is that we only use this contract to deploy a proxy contract. There should be no change in ABI between contracts deployed using both versions. Perhaps we could just make everything inherit from one and remove the other?

Downsides of that is that the bytecode won't match but the deployed contracts are already verified on mainnet and we can still interact with them using the ABI (and there seems to be no storage layout change either), so it might be okay to do it. Let me know if I'm overlooking something here.

> Replace OUSD-governance Timelock with TimelockGovernance

Sounds good

**Comment by sparrowDom** [comment]
I think you could put them all in `/root/contracts/contracts/governance` directory and move the contracts that are currently in that directory to `/root/contracts/contracts/ousd-governance`

**Comment by sparrowDom** [comment]
Oh I see you've done it the other way around. That is fine as well IMO

**Comment by sparrowDom** [comment]
Mental note: we will need to run a diff of the contract between 2 repos. Just to confirm there have not been any unintended changes

**Comment by sparrowDom** [comment]
> @shahthepro I think we can go ahead and use just one InitializeGovernedUpgradeabilityProxy, even though the implementation of isAddress function will be changed. (We could use the 4.6.0 Address version). The ABI won't have changes, so we will be able to interact with them. I don't see any issues with that, but will wait for someone else to confirm as well.

Yes go ahead and consolidate that into one contract. Storage slots are not impacted and I think we can use the 4.6.0 address with the [new way of identifying](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Address.sol#L41) if an address is a contract. And bump the pragma solidity compiler to ^0.8.1. for that file.
